src/tileset.rs

```diff
@@ -126,145 +102,170 @@ impl Tileset {
+            first_gid,
+            result_type: EmbeddedParseResultType::Embedded { tileset },
+        })
     }
 
+    /// Returns both the tileset and its first gid in the corresponding map.
     fn parse_xml_reference(
         attrs: &Vec<OwnedAttribute>,
-        path_relative_to: Option<&Path>,
-    ) -> Result<Tileset, TiledError> {
+        map_path: &Path,
+    ) -> Result<EmbeddedParseResult, TiledError> {
         let ((), (first_gid, source)) = get_attrs!(
             attrs,
             optionals: [],
             required: [
-                ("firstgid", first_gid, |v:String| v.parse().ok()),
+                ("firstgid", first_gid, |v:String| v.parse().ok().map(|n| Gid(n))),
                 ("source", name, |v| Some(v)),
             ],
             TiledError::MalformedAttributes("Tileset reference must have a firstgid and source with correct types".to_string())
         );
 
-        let tileset_path = path_relative_to
-            .ok_or(TiledError::SourceRequired {
-                object_to_parse: "Tileset".to_string(),
-            })?
-            .join(source);
-        let file = File::open(&tileset_path).map_err(|_| {
-            TiledError::Other(format!(
-                "External tileset file not found: {:?}",
-                tileset_path
-            ))
-        })?;
-        Tileset::new_external(file, first_gid, Some(&tileset_path))
+        let dir_containing_map = map_path.parent().ok_or(TiledError::PathIsNotFile)?;
+        let tileset_path = dir_containing_map.join(source);
+
+        Ok(EmbeddedParseResult {
+            first_gid,
+            result_type: EmbeddedParseResultType::ExternalReference { tileset_path },
+        })
     }
 
-    fn parse_external_tileset<R: Read>(
-        first_gid: u32,
-        parser: &mut EventReader<R>,
+    fn parse_external_tileset(
+        parser: &mut impl Iterator<Item = XmlEventResult>,
         attrs: &Vec<OwnedAttribute>,
-        path: Option<&Path>,
+        path: &Path,
     ) -> Result<Tileset, TiledError> {
-        let ((spacing, margin, tilecount, columns), (name, tile_width, tile_height)) = get_attrs!(
+        let ((spacing, margin, columns), (tilecount, name, tile_width, tile_height)) = get_attrs!(
             attrs,
             optionals: [
                 ("spacing", spacing, |v:String| v.parse().ok()),
                 ("margin", margin, |v:String| v.parse().ok()),
-                ("tilecount", tilecount, |v:String| v.parse().ok()),
                 ("columns", columns, |v:String| v.parse().ok()),
             ],
             required: [
+                ("tilecount", tilecount, |v:String| v.parse().ok()),
                 ("name", name, |v| Some(v)),
                 ("tilewidth", width, |v:String| v.parse().ok()),
                 ("tileheight", height, |v:String| v.parse().ok()),
             ],
             TiledError::MalformedAttributes("tileset must have a firstgid, name tile width and height with correct types".to_string())
```

Comment:
Should probably mention `tilecount` now, and should not mention `firstgid`.

And actually I don't see why `name` should be required, it can just default to an empty string.

Reply:
The `name` requirement is a separate issue; I'll fix the message though.